src/txn/txn_timestamp.c

```diff
@@ -1104,3 +1105,49 @@ __wt_txn_clear_read_timestamp(WT_SESSION_IMPL *session)
     }
     txn_shared->read_timestamp = WT_TS_NONE;
 }
+
+/*
+ * __wt_txn_checkpoint_cannot_start --
+ *     Return true if there's a transaction we need to wait for. This means transactions that have
+ *     begun committing with durable timestamp at or before the checkpoint timestamp, but have not
+ *     yet finished.
+ */
+bool
+__wt_txn_checkpoint_cannot_start(WT_SESSION_IMPL *session)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_TXN_GLOBAL *txn_global;
+    WT_TXN_SHARED *txn_shared;
+    wt_timestamp_t durable_ts;
+    uint32_t i, session_count;
+
+    conn = S2C(session);
+    txn_global = &conn->txn_global;
+
+    /* We're going to scan the table: wait for the lock. */
+    __wt_readlock(session, &txn_global->rwlock);
+
+    /* Walk the array of concurrent transactions. */
+    WT_ORDERED_READ(session_count, conn->session_cnt);
+    WT_STAT_CONN_INCR(session, txn_walk_sessions);
+    for (i = 0, txn_shared = txn_global->txn_shared_list; i < session_count; i++, txn_shared++) {
+        WT_STAT_CONN_INCR(session, txn_sessions_walked);
+
+        /*
+         * FUTURE: there is currently no way to tell if a transaction has started committing, or has
+         * only been assigned a durable or commit timestamp with timestamp_transaction(). It would
+         * be better not to wait for transactions that haven't actually started committing yet.
+         */
+        __txn_get_durable_timestamp(txn_shared, &durable_ts);
+
+        if (txn_shared->id == WT_TXN_NONE)
```

Comment:
How about performing this check first and then getting the `durable_ts`. Otherwise, we discard the `durable_ts` that were fetched.

Reply:
Good point; fixed in [8712dd0](https://github.com/wiredtiger/wiredtiger/pull/7774/commits/8712dd0231e759c54180235b2d1198b6fdc655f1).